nncf/openvino/quantization/weights_compression.py

```diff
@@ -9,23 +9,326 @@
+        rows.append(
+            [
+                bitwidth,
+                _proportion_str(n_total, total_num_weights, num_params),
+                _proportion_str(n_internal, num_internal_weights, num_internal_params),
+            ]
+        )
+
+    table = create_table(header, rows)
+    pretty_string = f"Statistics of the bitwidth distribution:\n{table}"
+    return pretty_string
+
+
+def _assign_mixed_precision(all_weight_params: List[WeightNodeParams], ratio: float, group_size: int) -> None:
+    """
+    Assigns mixed quantization scheme (e.g. uniform int8 or non-uniform nf4) for weights based on some criteria.
+
+    :param all_weight_params: List of information about each weight node. The quantization scheme is added to this info.
+    :param ratio: ratio between primary precision and backup (e.g. 0.9 means 90% of layers in NF4 and the rest in INT8).
+    :param group_size: number of weights (e.g. 64 or 128) that are independently quantized or share compression
+        parameters (scale). When it equals -1, per-channel quantization is assumed with group size equals to the size of
+        channel. Defaults to -1.
+    """
+    nf4_config = WeightCompressionConfig(num_bits=4, is_nf4=True, group_size=group_size)
+    if ratio != 1:
+        # NOTE: first and last layer is always in 8 bit.
+        errors = []
+        num_internal_weights = 0
+        for weight_param in all_weight_params[1:-1]:
+            weight = get_const_value(weight_param.weight_node)
+            axes = weight_param.reduction_axes
+            nf4_error = _get_nf4_error(weight, axes)
+            int8_error = _get_int8_err(weight, axes)
+            eps = np.finfo(weight.dtype).eps
+            error = nf4_error / (int8_error + eps)
+            errors.append(error)
+            num_internal_weights += weight_param.num_weights
+        # NOTE: index is defined in the array of all weight params by taking into account that errors were not
+        # calculated for first and last layers.
+        indexes_of_layers_in_ascending_order_of_errors = [
+            i[0] + 1 for i in sorted(enumerate(errors), reverse=False, key=lambda x: x[1])
+        ]
+        num_weights_in_4bit = 0
+        for index in indexes_of_layers_in_ascending_order_of_errors:
+            weight_param = all_weight_params[index]
+            current_ratio = (num_weights_in_4bit + weight_param.num_weights) / num_internal_weights
+            if current_ratio >= ratio:
+                break
+            weight_param.compression_config = nf4_config
+            num_weights_in_4bit += weight_param.num_weights
+
+    else:
+        for weight_param in all_weight_params[1:-1]:
+            weight_param.compression_config = nf4_config
+    nncf_logger.info(_get_bitwidth_distribution_str(all_weight_params))
+
+
+def insert_pre_compression_operations(
+    model: ov.Model,
+    mode: CompressWeightsMode,
+    ratio: float,
+    group_size: int,
```

Comment:
Docs for `ration` and `group_size` are missed

Reply:
thanks, added

Reply:
Sorry, I missed mode param as well :( 
And docstring for bits is outdated

Reply:
thanks, corrected